Library/ViewModels/ShippingRulesViewModel.swift

```diff
@@ -32,70 +33,73 @@ public protocol ShippingRulesViewModelType {
-    )
-    .map { data, newIndex in (data.1.firstIndex(of: data.2), newIndex) }
-    .filter { oldIndex, newIndex in oldIndex != newIndex }
-    .map(second)
+    let searchText = self.searchTextDidChangeProperty.signal
+      .ksr_debounce(.milliseconds(100), on: AppEnvironment.current.scheduler)
+      .skipNil()
 
-    let dataSelected = Signal.combineLatest(
-      dataInitial,
-      selectedIndex
-    )
-    .map { data, index in (data.0, data.1, index) }
-    .filter { _, shippingRules, index in index >= 0 && index < shippingRules.count }
-    .map { project, shippingRules, index in (project, shippingRules, shippingRules[index]) }
+    let selectedIndex = self.didSelectShippingRuleAtIndexProperty.signal
+      .skipNil()
 
-    let reloadDataWithShippingRulesInitial = dataInitial
-      .map(shippingRuleData(for:shippingRules:selectedShippingRule:))
-      .map { ($0, true) }
+    self.deselectCellAtIndex = selectedIndex
 
-    let reloadDataWithShippingRulesSelected = dataSelected
-      .map(shippingRuleData(for:shippingRules:selectedShippingRule:))
-      .map { ($0, false) }
+    let filteredData = initialData
+      .takePairWhen(searchText)
+      .map(dataMatching(_:searchText:))
 
-    self.reloadDataWithShippingRules = Signal.merge(
-      reloadDataWithShippingRulesInitial,
-      reloadDataWithShippingRulesSelected
+    let selectedShippingRule = Signal.merge(
+      initialData,
+      filteredData
     )
+    .takePairWhen(selectedIndex)
+    .map { data, selectedIndex in (data.1, selectedIndex) }
+    .map { shippingRules, selectedIndex in shippingRules[selectedIndex] }
 
-    self.flashScrollIndicators = reloadDataWithShippingRulesInitial
-      .ignoreValues()
-
-    let selectedShippingRuleIndexInitial = dataInitial
-      .map { _, shippingRules, selectedShippingRule in shippingRules.firstIndex(of: selectedShippingRule) }
-      .skipNil()
+    self.notifyDelegateOfSelectedShippingRule = selectedShippingRule
+      .skipRepeats()
 
-    let selectedShippingRuleIndexSelected = self.didSelectShippingRuleAtIndexProperty.signal
-      .skipNil()
+    let reloadDataInitial = initialData
+      .map(shippingRulesData(project:shippingRules:selectedShippingRule:))
+      .map { ($0, true) }
 
-    self.notifyDelegateOfSelectedShippingRule = dataSelected
-      .map(third)
+    let reloadDataFiltered = filteredData
+      .withLatest(from: Signal.merge(initialData.map(third), selectedShippingRule))
+      .map { data, selectedShippingRule in (data.0, data.1, selectedShippingRule) }
+      .map(shippingRulesData(project:shippingRules:selectedShippingRule:))
+      .map { ($0, true) }
 
-    self.scrollToCellAtIndex = Signal.combineLatest(
-      selectedShippingRuleIndexInitial,
-      reloadDataWithShippingRulesInitial
+    let reloadDataSelected = Signal.merge(
+      initialData,
+      filteredData
     )
-    .map(first)
+    .takePairWhen(selectedIndex)
+    .map { data, selectedIndex in (data.0, data.1, selectedIndex) }
```

Comment:
It's a little hard to follow what's happening when the tuples elements aren't named. It might be helpful to `typealias` the tuple as `ShippingRulesData` with named elements so this would read:

`.map { data, selectedIndex in (data.project, data.shippingRules, selectedIndex) }`